### DISCUSSION TITLE
Question @ multiple steppers

### DISCUSSION THREAD

**Discussion by dbtreasure** [Chr157i4n/PyTmcStepper#32] [Q&A] [answered]
_Thank you so much for developing this and putting it up on github, it's really interesting to look at and helps those of us who normally have to rely on circuit python to run drivers -> and therefore needing to look for another library to run this great driver_

I wanted to ask you if you could help elaborate what are the limiting factors to running mulitple steppers simultaneously, and is there any possible work around you've considered? I understand there is only a single UART pair of pins on the pi GPIO so would a second UART pair of pins make the pi usable if that were somehow possible? Or is the limiting factor transmitting messages to two UART devices? I've never run a multi UART device otherwise in my experience with Arduino and CircuitPython. 

I think my plan as-is is to utilize an Arduino with the AccelStepper library and then just send serial commands over a movement API for the robot I'm building. I want to use the Raspberry Pi to make integrating a camera and running some inferencing easier and had initially planned on having it directly drive the motors for the wheels but given there doesn't seem to be a feasible way I will probably just solve this using a microcontrollers capable of communicating with the pi over serial. 

Thanks so much for your time and consideration, and again thanks for publishing this great library. It's really educational reading through it.

**Comment by Chr157i4n** [comment, marked as answer]
Hello and thanks for the feedback,

running multiple drivers/motors is possible depending on your plans.

Please also have a look at:
[test_script_06_multiple_drivers.py](https://github.com/Chr157i4n/TMC2209_Raspberry_Pi/blob/main/tests/test_script_06_multiple_drivers.py)
[Can I use two TMC2209 with one Raspberry Pi board with this lib? #3 ](https://github.com/Chr157i4n/TMC2209_Raspberry_Pi/issues/3)

There are two things to consider:

### 1. UART-Interface
this is described in chapter _4.3 UART Signals_ in the [datasheet](https://www.trinamic.com/fileadmin/assets/Products/ICs_Documents/TMC2209_datasheet_rev1.08.pdf).
if you have <= 4 drivers, you can use the built in addressing of the driver trough the MS1_ADDR0 and MS2_ADDR1 pins.
![grafik](https://user-images.githubusercontent.com/8201379/209454014-14fa3a44-5c54-4c08-9c28-27ea22ffa9ef.png)
You can use this to communicate with up to 4 drivers via UART with only one UART Interface on the Raspberry Pi. This is also already implemented in this library.
If you have more drivers, you either need more UART-Interfaces or you need a switch in the UART Line, like this:
![grafik](https://user-images.githubusercontent.com/8201379/209454059-7520e697-ad44-48a5-9b81-ba30782d6a8d.png)
Via UART you can move the motor, but the TMC2209 only supports setting a velocity with VACTUAL (no accurate positioning). For accurate positiong you need a different driver (like the TMC5160) or use the STEP/DIR-Interface.

### 2. STEP/DIR-Interface
To accuractly position a driver/motor you need to use the STEP/DIR-Interface on the TMC2209. If you want to run multiple driver/motor, each driver needs to have their own STEP/DIR-Interface (2 GPIO-Pins) on the Raspberry Pi. The Problem here is to run the motors simultaneously.
Currently the [run function](https://github.com/Chr157i4n/TMC2209_Raspberry_Pi/blob/457dcf27a581f8796ac30cd43ada20e70a577b07/src/TMC_2209/TMC_2209_StepperDriver.py#L1103-L1108) blocks the code until it is finished. So currently to archieve simultaneous movement the run function for each driver needs to run on its own threads.

I hope this helps. If you have further questions feel free to ask ;)

**Comment by dbtreasure** [comment]
Hey @Chr157i4n thanks so much for this answer, my apologies for not checking the *Issues* section of this library so thank you for restating some things.

**Comment by probably-Erwins-Cat** [comment]
in case a RPi is using multiple steppers, are multiple DIAG traces to separate GPIOs needed, or can one daisy-chained DIAG trace be used?

**Reply by Chr157i4n** [reply]
So you want to use multiple steppers with StallGuards?
The pin just gives a high signal on stall detection:
![grafik](https://github.com/user-attachments/assets/2a974460-68c3-4c2c-bec5-3a9561bda9dd)
If you would connect several DIAG outputs of the TMC2209 to one input pin of the RPi, you propably create voltage divider when one driver is signaling a stall and the others are not. I am not sure if this will work at all (it depends on the resistors on the TMC2209).

Even if it works, you can no longer tell which stepper is stalling.

So i would suggest using seperate Pins for each DIAG output.
